File: tests/Tests.CSharp/UncertainClockCrossVerifyTests.cs
using System.IO;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp;
using Hlc = Zeta.Core.CSharp.UncertainClock.Hlc;

namespace Zeta.Tests.CSharp;

/// <summary>
/// UncertainClock cross-language agreement — the C# oracle replays the shared seed
/// (<c>src/Core.TypeScript/uncertain-clock/golden-vectors.json</c>) that the F#/TS/Rust oracles also
/// verify. All values are exact int64, so the full surface byte-locks (no float caveat).
/// </summary>
public class UncertainClockCrossVerifyTests
{
    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(UncertainClockCrossVerifyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
        {
            dir = dir.Parent;
        }

        return dir?.FullName ?? throw new DirectoryNotFoundException("Could not locate repo root (Zeta.sln).");
    }

    private static JsonElement Seed()
    {
        var path = Path.Join(RepoRoot(), "src", "Core.TypeScript", "uncertain-clock", "golden-vectors.json");
        Assert.True(File.Exists(path), $"seed not found: {path}");
        using var doc = JsonDocument.Parse(File.ReadAllText(path));
        return doc.RootElement.Clone();
    }

    private static Hlc ReadHlc(JsonElement e) =>
        new(e.GetProperty("physical").GetInt64(), e.GetProperty("logical").GetInt64());

    [Fact]
    public void CompareHlcAgreesWithSeed()
    {
        foreach (var v in Seed().GetProperty("compareHlc").EnumerateArray())
        {
            Assert.Equal(v.GetProperty("result").GetInt32(),
                UncertainClock.CompareHlc(ReadHlc(v.GetProperty("a")), ReadHlc(v.GetProperty("b"))));
        }
    }

    [Fact]
    public void SendAgreesWithSeed()
    {
        foreach (var v in Seed().GetProperty("send").EnumerateArray())
        {
            Assert.Equal(ReadHlc(v.GetProperty("result")),
                UncertainClock.Send(ReadHlc(v.GetProperty("clock")), v.GetProperty("now").GetInt64()));
        }
    }

    [Fact]
    public void ReceiveAgreesWithSeed()
    {
        foreach (var v in Seed().GetProperty("receive").EnumerateArray())
        {
            Assert.Equal(ReadHlc(v.GetProperty("result")),
                UncertainClock.Receive(ReadHlc(v.GetProperty("clock")), ReadHlc(v.GetProperty("msg")), v.GetProperty("now").GetInt64()));
        }
    }

    [Fact]
    public void DefinitelyBeforeAgreesWithSeed()
    {
        foreach (var v in Seed().GetProperty("definitelyBefore").EnumerateArray())
        {
            var a = v.GetProperty("a");
            var b = v.GetProperty("b");
            Assert.Equal(v.GetProperty("result").GetBoolean(),
                UncertainClock.DefinitelyBefore(
                    a.GetProperty("physical").GetInt64(), a.GetProperty("eps").GetInt64(),
                    b.GetProperty("physical").GetInt64(), b.GetProperty("eps").GetInt64()));
        }
    }

    [Fact]
    public void UncertainAgreesWithSeed()
    {
        foreach (var v in Seed().GetProperty("uncertain").EnumerateArray())
        {
            var a = v.GetProperty("a");
            var b = v.GetProperty("b");
            Assert.Equal(v.GetProperty("result").GetBoolean(),
                UncertainClock.Uncertain(
                    a.GetProperty("physical").GetInt64(), a.GetProperty("eps").GetInt64(),
                    b.GetProperty("physical").GetInt64(), b.GetProperty("eps").GetInt64()));
        }
    }
}

File: tests/Tests.FSharp/UncertainClock.CrossVerify.Tests.fs
module Zeta.Tests.UncertainClockCrossVerifyTests

open System.IO
open System.Text.Json
open global.Xunit
open Zeta.Core

module UC = Zeta.Core.UncertainClock

// ═══════════════════════════════════════════════════════════════════
// UncertainClock cross-language agreement — the F# oracle uses the REAL canonical UncertainClock and
// replays the shared seed (src/Core.TypeScript/uncertain-clock/golden-vectors.json) that the C#/TS/Rust
// oracles also verify. Every value is exact int64, so the full surface byte-locks (no float caveat):
// compareHlc, send, receive, definitelyBefore, uncertain.
// ═══════════════════════════════════════════════════════════════════

let private repoRoot () =
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then failwith "Could not locate repo root (Zeta.sln)." else dir.FullName

let private hlc (e: JsonElement) : UC.Hlc =
    { Physical = (e.GetProperty "physical").GetInt64(); Logical = (e.GetProperty "logical").GetInt64() }

[<Fact>]
let ``F# UncertainClock agrees with the shared golden seed`` () =
    let path = Path.Join(repoRoot (), "src", "Core.TypeScript", "uncertain-clock", "golden-vectors.json")
    Assert.True(File.Exists path, sprintf "seed not found: %s" path)
    use doc = JsonDocument.Parse(File.ReadAllText path)
    let root = doc.RootElement
    let section (name: string) : JsonElement[] = root.GetProperty(name).EnumerateArray() |> Seq.toArray

    for v in section "compareHlc" do
        Assert.Equal<int>((v.GetProperty "result").GetInt32(), UC.compareHlc (hlc (v.GetProperty "a")) (hlc (v.GetProperty "b")))

    for v in section "send" do
        Assert.Equal<UC.Hlc>(hlc (v.GetProperty "result"), UC.send (hlc (v.GetProperty "clock")) ((v.GetProperty "now").GetInt64()))

    for v in section "receive" do
        Assert.Equal<UC.Hlc>(hlc (v.GetProperty "result"), UC.receive (hlc (v.GetProperty "clock")) (hlc (v.GetProperty "msg")) ((v.GetProperty "now").GetInt64()))

    let mk (e: JsonElement) : UC.Uncertain =
        UC.make { Physical = (e.GetProperty "physical").GetInt64(); Logical = 0L } ((e.GetProperty "eps").GetInt64())

    for v in section "definitelyBefore" do
        Assert.Equal<bool>((v.GetProperty "result").GetBoolean(), UC.definitelyBefore (mk (v.GetProperty "a")) (mk (v.GetProperty "b")))

    for v in section "uncertain" do
        Assert.Equal<bool>((v.GetProperty "result").GetBoolean(), UC.uncertain (mk (v.GetProperty "a")) (mk (v.GetProperty "b")))
